rmk/src/controller/mod.rs

```diff
@@ -0,0 +1,173 @@
+//! Controller module for RMK
+//!
+//! This module defines the `Controller` trait and several macros for running output device controllers.
+//! The `Controller` trait provides the interface for individual output device controllers, and the macros facilitate their concurrent execution.
+
+use embassy_futures::select::{select, Either};
+
+/// Common trait for controllers.
+pub trait Controller {
+    /// Type of the received events
+    type Event;
+
+    /// Block waiting for next message
+    async fn next_message(&mut self) -> Self::Event;
+}
+
+/// The trait for event-driven controllers.
+///
+/// This trait defines the interface for event-driven controllers in RMK.
+///
+/// # Example
+/// ```rust
+/// // Define a controller
+/// struct MyController;
+///
+/// impl Controller for MyController { /* ... */ }
+///
+/// impl EventController for MyController {
+///     async fn process_event(&mut self, event: Self::Event) {
+///         // handle event
+///     }
+/// }
+///
+/// // Use the input device
+/// let c = MyController;
+///
+/// // Run device simultaneously with RMK
+/// embassy_futures::join::join(
+///     c.event_loop(),
+///     run_rmk(
+///         // ...
+///     ),
+/// )
+/// .await;
+/// ```
+pub trait EventController: Controller {
+    /// Process the received event
+    async fn process_event(&mut self, event: Self::Event);
+
+    /// Event loop
+    async fn event_loop(&mut self) {
+        loop {
+            let event = self.next_message().await;
+            self.process_event(event).await;
+        }
+    }
+}
+
+/// The trait for polling controllers.
+///
+/// This trait defines the interface for polling controllers in RMK.
+///
+/// # Example
+/// ```rust
+/// // Define a controller
+/// struct MyController;
+///
+/// impl Controller for MyController { /* ... */ }
+///
+/// impl PollingController for MyController {
+///     type INTERVAL: embassy_time::Duration = embassy_time::Duration::from_hz(60);
+///
+///     async fn process_event(&mut self, event: Self::Event) {
+///         // handle event
+///     }
+///
+///     async fn update(&mut self) {
+///         // update periodic
+///     }
+/// }
+///
+/// // Use the input device
+/// let c = MyController;
+///
+/// // Run device simultaneously with RMK
+/// embassy_futures::join::join(
+///     c.polling_loop(),
+///     run_rmk(
+///         // ...
+///     ),
+/// )
+/// .await;
+/// ```
+pub trait PollingController: Controller {
+    /// Interval between `update` calls
+    const INTERVAL: embassy_time::Duration;
+
+    /// Process the received event
+    async fn process_event(&mut self, event: Self::Event);
```

Comment:
I prefer to move `process_event` to `Controller`, even though they're usually differently implemented in `PollingController` and `EventController` -- that's why there're `PollingController` and `EventController` based on the `Controller`.

Reply:
Should `EventController` and `PollingController` be mutually exclusive then?

Reply:
I think so